explanation/performance/perf-pgo.md

```diff
@@ -0,0 +1,153 @@
+```
+
+### Building OpenSSL for profiling
+
+Before we are able to profile OpenSSL, we need sure that we compile the software in a way that the generated program meets the requirements.  In our case, you can use the `file` command after building the binary to make sure that it has not been stripped:
+
+```bash
+$ file /usr/bin/openssl
+/usr/bin/openssl: ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=42a5ac797b6981bd45a9ece3f67646edded56bdb, for GNU/Linux 3.2.0, with debug_info, not stripped
+```
+
+You can use the `eu-readelf` command (from the `elfutils` package) to determine whether any compression was performed with `dwz`:
+
+```bash
+$ eu-readelf -S /usr/bin/openssl
+[...]
+Section Headers:
+[Nr] Name                 Type         Addr             Off      Size     ES Flags Lk Inf Al
+[...]
+[31] .debug_abbrev        PROGBITS     0000000000000000 001bcbf3 000110c4  0        0   0  1
+[...]
+```
+
+If the `Flags` field has `C` in it, this means that `dwz` was used.  You will then need to figure out how to disable its execution during the build.
+
+### Using `perf` to obtain profile data
+
+With OpenSSL prepared and ready to be profiled, it's now time to use `perf` to obtain profiling data from our workload.  If you are on an Intel system, you can invoke `perf` using:
+
+```bash
+$ sudo perf record \
+	-e br_inst_retired.near_taken \
+	--branch-any \
+	--all-cpus \
+	--output openssl-nonpgo.perfdata \
+	-- openssl speed -mr -seconds 60 -evp md5 sha512 rsa2048 aes-256-cbc
+```
+
+On an AMD system, use the following invocation:
+
+```bash
+$ sudo perf record \
+	-e 'cpu/event=0xc4,umask=0x0,name=ex_ret_brn_tkn/' \
+	--branch-any \
+	--all-cpus \
+	--output openssl-nonpgo.perfdata \
+	-- openssl speed -mr -seconds 60 -evp md5 sha512 rsa2048 aes-256-cbc
+```
+
+Note that the only thing that differs is the PMU event to be monitored.
+
+After the command has finished running, you should see a file named `openssl-nonpgo.perfdata` in your current directory.  This is the file we will now need to convert to `gcov`.
+
+### Converting `perfdata` to `gcov` with `autofdo`
+
+With `autofdo` installed, you can convert the `perfdata` generated in the last step to `gcov` by doing:
+
+```bash
+$ sudo chown ubuntu:ubuntu openssl-nonpgo.perfdata
+$ create_gcov \
+	--binary /usr/bin/openssl \
+	--gcov openssl-nonpgo.gcov \
+	--profile openssl-nonpgo.perfdata \
+	--gcov_version 2 \
+	-use_lbr false
+```
+
+`create_gcov` is verbose and will display several messages that may look like something is wrong; you should inspect its output and also make sure to check the `$?` shell variable to make sure that it has finished successfully.  You will then have a file called `openssl-nonpgo.gcov` in your current directory, which is the `gcov` file we will use to feed back to GCC when recompiling OpenSSL.  You can inspect the file by using the `dump_gcov` tool.
```

Comment:
"several messages that may look like something is wrong"

can we assist the reader with examples what is "looking like but not bad" and what is "really bad"?

Reply:
Examples are always good :)

Reply:
Sure thing.  I put an example of what a successful invocation looks like.